fvm/evm/emulator/state/stateDB.go

```diff
@@ -0,0 +1,462 @@
+	return db.lastestView().SlotInAccessList(types.SlotAddress{Address: addr, Key: key})
+}
+
+// AddAddressToAccessList adds the given address to the access list.
+func (db *StateDB) AddAddressToAccessList(addr gethCommon.Address) {
+	db.lastestView().AddAddressToAccessList(addr)
+}
+
+// AddSlotToAccessList adds the given (address,slot) to the access list.
+func (db *StateDB) AddSlotToAccessList(addr gethCommon.Address, key gethCommon.Hash) {
+	db.lastestView().AddSlotToAccessList(types.SlotAddress{Address: addr, Key: key})
+}
+
+// AddLog appends a lot to the collection of logs
+func (db *StateDB) AddLog(log *gethTypes.Log) {
+	db.lastestView().AddLog(log)
+}
+
+// AddPreimage adds a preimage to the collection of preimages
+func (db *StateDB) AddPreimage(hash gethCommon.Hash, data []byte) {
+	db.lastestView().AddPreimage(hash, data)
+}
+
+// RevertToSnapshot reverts the changes until we reach the given snaptshot
+func (db *StateDB) RevertToSnapshot(index int) {
+	if index > len(db.views) {
+		db.cachedError = fmt.Errorf("invalid revert")
+		return
+	}
+	db.views = db.views[:index]
+}
+
+// Snapshot takes an snapshot of the state and returns an int
+// that can be used later for revert calls.
+func (db *StateDB) Snapshot() int {
+	newView := db.lastestView().NewChildView()
+	db.views = append(db.views, newView)
+	return len(db.views) - 1
+}
+
+// Logs returns the list of logs
+// it also update each log with the block and tx info
+func (db *StateDB) Logs(
+	blockHash gethCommon.Hash,
+	blockNumber uint64,
+	txHash gethCommon.Hash,
+	txIndex uint,
+) []*gethTypes.Log {
+	allLogs := make([]*gethTypes.Log, 0)
+	for _, view := range db.views {
+		for _, log := range view.Logs() {
+			log.BlockNumber = blockNumber
+			log.BlockHash = blockHash
+			log.TxHash = txHash
+			log.TxIndex = txIndex
+			allLogs = append(allLogs, log)
+		}
+	}
+	return allLogs
+}
+
+// Preimages returns a set of preimages
+func (db *StateDB) Preimages() map[gethCommon.Hash][]byte {
+	preImages := make(map[gethCommon.Hash][]byte, 0)
+	for _, view := range db.views {
+		for k, v := range view.Preimages() {
+			preImages[k] = v
+		}
+	}
+	return preImages
+}
+
+// Commit commits state changes back to the underlying
+func (db *StateDB) Commit() error {
+	// return error if any has been acumulated
+	if db.cachedError != nil {
+		return wrapError(db.cachedError)
+	}
+
+	var err error
+
+	// iterate views and collect dirty addresses and slots
+	addresses := make(map[gethCommon.Address]interface{})
+	slots := make(map[types.SlotAddress]interface{})
+	for _, view := range db.views {
+		for key := range view.DirtyAddresses() {
+			addresses[key] = struct{}{}
+		}
+		for key := range view.DirtySlots() {
+			slots[key] = struct{}{}
+		}
+	}
+
+	// sort addresses
+	sortedAddresses := make([]gethCommon.Address, 0, len(addresses))
+	for addr := range addresses {
+		sortedAddresses = append(sortedAddresses, addr)
+	}
+	sort.Slice(sortedAddresses,
```

Comment:
nit? maybe better use stable sort with generics:
```go
slices.SortFunc(sortedAddresses, func(a, b gethCommon.Address) int {
		return bytes.Compare(a[:], b[:])
	})
```

Reply:
updated

Reply:
a more modern API that uses generics and does stable sort, which here might not really be important, that's why I wrote nit, feel free to ignore my nit comments 😆

Reply:
I reverted some of these changes, since the Geth version was not happy about it, I'll put them back when I update the Geth version